src/ledger/LedgerStateImpl.h

```diff
@@ -29,156 +29,260 @@ class LedgerState::Impl
     void checkNotSealed() const;
 
+    // getDeltaVotes has the basic exception safety guarantee. If it throws an
+    // exception other than LedgerStateRootFatalError, then
+    // - the prepared statement cache may be, but is not guaranteed to be,
+    //   modified
+    // - the entry cache may be, but is not guaranteed to be, cleared.
+    // If it throws LedgerStateRootFatalError, there are no guarantees.
     std::map<AccountID, int64_t> getDeltaVotes() const;
 
+    // getTotalVotes has the strong exception safety guarantee
     std::map<AccountID, int64_t>
     getTotalVotes(std::vector<InflationWinner> const& parentWinners,
                   std::map<AccountID, int64_t> const& deltaVotes,
                   int64_t minVotes) const;
 
+    // enumerateInflationWinners has the strong exception safety guarantee
     std::vector<InflationWinner>
     enumerateInflationWinners(std::map<AccountID, int64_t> const& totalVotes,
                               size_t maxWinners, int64_t minVotes) const;
 
-    void sealAndMaybeUpdateLastModified();
+    // getEntryIterator has the strong exception safety guarantee
+    EntryIterator getEntryIterator(EntryMap const& entries) const;
+
+    // maybeUpdateLastModified has the strong exception safety guarantee
+    EntryMap maybeUpdateLastModified() const;
+
+    // maybeUpdateLastModifiedThenInvokeThenSeal has the same exception safety
+    // guarantee as f
+    void maybeUpdateLastModifiedThenInvokeThenSeal(
+        std::function<void(EntryMap const&)> f);
 
   public:
+    // Constructor has the strong exception safety guarantee
     Impl(LedgerState& self, AbstractLedgerStateParent& parent,
          bool shouldUpdateLastModified);
 
+    // addChild has the strong exception safety guarantee
     void addChild(AbstractLedgerState& child);
 
+    // commit has the strong exception safety guarantee. If
+    // LedgerStateRootFatal error is thrown, there are no guarantees.
     void commit(Identifier id);
 
-    void commitChild();
+    // commitChild has the strong exception safety guarantee. If
+    // LedgerStateRootFatal error is thrown, there are no guarantees.
+    void commitChild(EntryIterator iter);
 
+    // create has the basic exception safety guarantee. If it throws an
+    // exception other than LedgerStateRootFatalError, then
+    // - the prepared statement cache may be, but is not guaranteed to be,
+    //   modified
+    // - the entry cache may be, but is not guaranteed to be, cleared.
+    // If it throws LedgerStateRootFatalError, there are no guarantees.
     LedgerStateEntry create(LedgerState& self, LedgerEntry const& entry);
 
+    // deactivate has the strong exception safety guarantee
     void deactivate(LedgerKey const& key);
 
+    // deactivateHeader has the strong exception safety guarantee
     void deactivateHeader();
 
+    // erase has the basic exception safety guarantee. If it throws an exception
+    // other than LedgerStateRootFatalError, then
```

Comment:
comment applies to other methods:
as a caller, I am not sure what those guarantees really mean and which recovery steps I should (or can) take. `LedgerStateRootFatalError` at least is clear ("you're dead");
in particular: if a cache is out of sync - does that mean that I am hosed?

Reply:
Good point, need to clarify the language here. A guarantee like "the entry cache may be, but is not guaranteed to be, cleared" is intended to mean that you will get the same results but the exact data access pattern may be different (cache miss instead of cache hit on some future query) assuming that the change in data access pattern does not lead to some spurious exception (which could happen if a key was dropped from the cache due to a thrown exception and then a database query failed after the subsequent cache miss). The overall idea here was that functions that offers exception safety guarantees should be safe to continue using after a thrown exception _unless_ the thrown exception was `LedgerStateRootFatalError` in which case there is no way to proceed.